In [165]:
# Import pandas
import pandas as pd

In [166]:
# Upload 'bike_rentals.csv' to dataFrame
df_bikes = pd.read_csv('bike_rentals.csv')

In [167]:
# Get the shape (dimensions) of the DataFrame.
# Returns a tuple: (number_of_rows, number_of_columns)
# Useful for quickly checking how many records (rows) and features (columns) the dataset has.
df_bikes.shape

(731, 16)

In [168]:
# Display first 5 rows
df_bikes.head()

,instant,dteday,season,yr,mnth,holiday,weekday,workingday,weathersit,temp,atemp,hum,windspeed,casual,registered,cnt
0,1,2011-01-01,1.0,0.0,1.0,0.0,6.0,0.0,2,0.344167,0.363625,0.805833,0.160446,331,654,985
1,2,2011-01-02,1.0,0.0,1.0,0.0,0.0,0.0,2,0.363478,0.353739,0.696087,0.248539,131,670,801
2,3,2011-01-03,1.0,0.0,1.0,0.0,1.0,1.0,1,0.196364,0.189405,0.437273,0.248309,120,1229,1349
3,4,2011-01-04,1.0,0.0,1.0,0.0,2.0,1.0,1,0.200000,0.212122,0.590435,0.160296,108,1454,1562
4,5,2011-01-05,1.0,0.0,1.0,0.0,3.0,1.0,1,0.226957,0.229270,0.436957,0.186900,82,1518,1600


In [169]:
# Show df_bikes descriptive statistics
df_bikes.describe()

,instant,season,yr,mnth,holiday,weekday,workingday,weathersit,temp,atemp,hum,windspeed,casual,registered,cnt
count,731.000000,731.000000,730.000000,730.000000,731.000000,731.000000,731.000000,731.000000,730.000000,730.000000,728.000000,726.000000,731.000000,731.000000,731.000000
mean,366.000000,2.496580,0.500000,6.512329,0.028728,2.997264,0.682627,1.395349,0.495587,0.474512,0.627987,0.190476,848.176471,3656.172367,4504.348837
std,211.165812,1.110807,0.500343,3.448303,0.167155,2.004787,0.465773,0.544894,0.183094,0.163017,0.142331,0.077725,686.622488,1560.256377,1937.211452
min,1.000000,1.000000,0.000000,1.000000,0.000000,0.000000,0.000000,1.000000,0.059130,0.079070,0.000000,0.022392,2.000000,20.000000,22.000000
25%,183.500000,2.000000,0.000000,4.000000,0.000000,1.000000,0.000000,1.000000,0.336875,0.337794,0.521562,0.134494,315.500000,2497.000000,3152.000000
50%,366.000000,3.000000,0.500000,7.000000,0.000000,3.000000,1.000000,1.000000,0.499166,0.487364,0.627083,0.180971,713.000000,3662.000000,4548.000000
75%,548.500000,3.000000,1.000000,9.750000,0.000000,5.000000,1.000000,2.000000,0.655625,0.608916,0.730104,0.233218,1096.000000,4776.500000,5956.000000
max,731.000000,4.000000,1.000000,12.000000,1.000000,6.000000,1.000000,3.000000,0.861667,0.840896,0.972500,0.507463,3410.000000,6946.000000,8714.000000


## 📊 What Values Does `df.describe()` Return?

The `describe()` method in pandas provides summary statistics of the DataFrame’s numeric columns by default.

### Summary of Returned Values:

- **Count**  
  The number of **non-missing (non-NaN)** values in the column.

- **Mean**  
  The **average** value of the column.  
  Calculated as:
  $$
  \text{Mean} = \frac{\sum x_i}{n}
  $$
- **Standard Deviation**  
  A measure of how **spread out** the values are around the mean.  
  Calculated using **sample standard deviation**:  
$$
  \text{Std} = \sqrt{\frac{\sum (x_i - \bar{x})^2}{n - 1}}
$$

- **Minimum**  
  The **smallest** value in the column.

- **Maximum**  
  The **largest** value in the column.

- **Percentiles**  
  Values that divide the **ordered data** into 100 equal parts.  
  By default, `describe()` returns:
  - **25th percentile (Q1)** – lower quartile
  - **50th percentile (Q2)** – median
  - **75th percentile (Q3)** – upper quartile


In [170]:
# Display df_bikes info
df_bikes.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 731 entries, 0 to 730
Data columns (total 16 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   instant     731 non-null    int64  
 1   dteday      731 non-null    object 
 2   season      731 non-null    float64
 3   yr          730 non-null    float64
 4   mnth        730 non-null    float64
 5   holiday     731 non-null    float64
 6   weekday     731 non-null    float64
 7   workingday  731 non-null    float64
 8   weathersit  731 non-null    int64  
 9   temp        730 non-null    float64
 10  atemp       730 non-null    float64
 11  hum         728 non-null    float64
 12  windspeed   726 non-null    float64
 13  casual      731 non-null    int64  
 14  registered  731 non-null    int64  
 15  cnt         731 non-null    int64  
dtypes: float64(10), int64(5), object(1)
memory usage: 91.5+ KB


**Finding the number of null values**

In [171]:
# Sum null values across entire dataset
# The first sums the null values of each column, while the second method sums the column counts.
df_bikes.isna().sum().sum()

np.int64(12)

**Displaying null values**

In [172]:
# Show null value in df_bikes
# df_bikes[df_bikes.isna().any(axis=1)]
# Show only rows with at least one NaN, and highlight NaNs with red background
df_bikes[df_bikes.isna().any(axis=1)].style.highlight_null(color='red')

,instant,dteday,season,yr,mnth,holiday,weekday,workingday,weathersit,temp,atemp,hum,windspeed,casual,registered,cnt
56,57,2011-02-26,1.000000,0.000000,2.000000,0.000000,6.000000,0.000000,1,0.282500,0.282192,0.537917,nan,424,1545,1969
81,82,2011-03-23,2.000000,0.000000,3.000000,0.000000,3.000000,1.000000,2,0.346957,0.337939,0.839565,nan,203,1918,2121
128,129,2011-05-09,2.000000,0.000000,5.000000,0.000000,1.000000,1.000000,1,0.532500,0.525246,0.588750,nan,664,3698,4362
129,130,2011-05-10,2.000000,0.000000,5.000000,0.000000,2.000000,1.000000,1,0.532500,0.522721,nan,0.115671,694,4109,4803
213,214,2011-08-02,3.000000,0.000000,8.000000,0.000000,2.000000,1.000000,1,0.783333,0.707071,nan,0.205850,801,4044,4845
298,299,2011-10-26,4.000000,0.000000,10.000000,0.000000,3.000000,1.000000,2,0.484167,0.472846,0.720417,nan,404,3490,3894
388,389,2012-01-24,1.000000,1.000000,1.000000,0.000000,2.000000,1.000000,1,0.342500,0.349108,nan,0.123767,439,3900,4339
528,529,2012-06-12,2.000000,1.000000,6.000000,0.000000,2.000000,1.000000,2,0.653333,0.597875,0.833333,nan,477,4495,4972
701,702,2012-12-02,4.000000,1.000000,12.000000,0.000000,0.000000,0.000000,2,nan,nan,0.823333,0.124379,892,3757,4649
730,731,2012-12-31,1.000000,nan,nan,0.000000,1.000000,0.000000,2,0.215833,0.223487,0.577500,0.154846,439,2290,2729


# **Correcting null values**

**Replacing with the median/mean**

In [173]:
# Before correction
# Display rows 56 and 81
df_bikes.iloc[[56, 81]].style.highlight_null(color='red')

,instant,dteday,season,yr,mnth,holiday,weekday,workingday,weathersit,temp,atemp,hum,windspeed,casual,registered,cnt
56,57,2011-02-26,1.000000,0.000000,2.000000,0.000000,6.000000,0.000000,1,0.282500,0.282192,0.537917,nan,424,1545,1969
81,82,2011-03-23,2.000000,0.000000,3.000000,0.000000,3.000000,1.000000,2,0.346957,0.337939,0.839565,nan,203,1918,2121


In [174]:
# Fill windspeed null values with median
df_bikes['windspeed'] = df_bikes['windspeed'].fillna(df_bikes['windspeed'].median())

In [175]:
# After correction
# Display rows 56 and 81
median_windspeed = df_bikes['windspeed'].median()

def highlight_filled(val):
    color = 'lightgreen' if val == median_windspeed else ''
    return f'background-color: {color}'

df_bikes.iloc[[56, 81]].style.map(highlight_filled, subset=['windspeed'])

,instant,dteday,season,yr,mnth,holiday,weekday,workingday,weathersit,temp,atemp,hum,windspeed,casual,registered,cnt
56,57,2011-02-26,1.000000,0.000000,2.000000,0.000000,6.000000,0.000000,1,0.282500,0.282192,0.537917,0.180971,424,1545,1969
81,82,2011-03-23,2.000000,0.000000,3.000000,0.000000,3.000000,1.000000,2,0.346957,0.337939,0.839565,0.180971,203,1918,2121


## *TIP*
The median is often a better choice than the mean. The median guarantees that half the data is greater than the given value and half the data is lower. The mean, by contrast, is vulnerable to outliers.

**Groupby with the median/mean**

In [176]:
# Groupby season with median
# Group the DataFrame by the 'season' column.
# For each group (i.e., each unique season), calculate the median of all numeric columns.
# This gives a new DataFrame where each row corresponds to a season,
# and each column shows the median value for that numeric attribute within that season.
# df_bikes.groupby(['season']).median()
df_bikes.drop(columns=['dteday']).groupby('season').median()

,instant,yr,mnth,holiday,weekday,workingday,weathersit,temp,atemp,hum,windspeed,casual,registered,cnt
season,,,,,,,,,,,,,,
1.0,366.0,0.5,2.0,0.0,3.0,1.0,1.0,0.285833,0.282821,0.543750,0.202750,218.0,1867.0,2209.0
2.0,308.5,0.5,5.0,0.0,3.0,1.0,1.0,0.562083,0.538212,0.646667,0.191546,867.0,3844.0,4941.5
3.0,401.5,0.5,8.0,0.0,3.0,1.0,1.0,0.714583,0.656575,0.635833,0.165115,1050.5,4110.5,5353.5
4.0,493.0,0.5,11.0,0.0,3.0,1.0,1.0,0.410000,0.409708,0.661042,0.167918,544.5,3815.0,4634.5


In [177]:
# Before Nan values are transformed from hum column
df_bikes.iloc[[129, 213, 388]]

,instant,dteday,season,yr,mnth,holiday,weekday,workingday,weathersit,temp,atemp,hum,windspeed,casual,registered,cnt
129,130,2011-05-10,2.0,0.0,5.0,0.0,2.0,1.0,1,0.532500,0.522721,NaN,0.115671,694,4109,4803
213,214,2011-08-02,3.0,0.0,8.0,0.0,2.0,1.0,1,0.783333,0.707071,NaN,0.205850,801,4044,4845
388,389,2012-01-24,1.0,1.0,1.0,0.0,2.0,1.0,1,0.342500,0.349108,NaN,0.123767,439,3900,4339


In [178]:
# Convert 'hum' null values to median of season

# For each row in the DataFrame, fill missing values in the 'hum' (humidity) column
# using the median humidity value for that row's corresponding 'season'.

# Step-by-step breakdown:
# 1. df_bikes.groupby('season')['hum']: Groups the 'hum' column by 'season'.
# 2. .transform('median'): Calculates the median humidity for each season
#    and returns a Series the same length as the original DataFrame,
#    where each value is the median humidity for that row's season.
# 3. df_bikes['hum'].fillna(...): Fills NaN values in 'hum' with the season-specific medians.
# 4. The result is assigned back to df_bikes['hum'], updating the column with filled values.
df_bikes['hum'] = df_bikes['hum'].fillna(df_bikes.groupby('season')['hum'].transform('median'))

In [179]:
# Note: row #129 is in season #2, #213 is in season #3 and #388 is in season #1
df_bikes.iloc[[129, 213, 388]]

,instant,dteday,season,yr,mnth,holiday,weekday,workingday,weathersit,temp,atemp,hum,windspeed,casual,registered,cnt
129,130,2011-05-10,2.0,0.0,5.0,0.0,2.0,1.0,1,0.532500,0.522721,0.646667,0.115671,694,4109,4803
213,214,2011-08-02,3.0,0.0,8.0,0.0,2.0,1.0,1,0.783333,0.707071,0.635833,0.205850,801,4044,4845
388,389,2012-01-24,1.0,1.0,1.0,0.0,2.0,1.0,1,0.342500,0.349108,0.543750,0.123767,439,3900,4339


**Obtaining the median/mean from specific rows**

In some cases, it may be advantageous to replace null values with data from specific rows.

When correcting temperature, aside from consulting historical records, taking the mean temperature of the day before and the day after should give a good estimate.

In [180]:
# Show null values of 'temp' column
df_bikes[df_bikes['temp'].isna()]

,instant,dteday,season,yr,mnth,holiday,weekday,workingday,weathersit,temp,atemp,hum,windspeed,casual,registered,cnt
701,702,2012-12-02,4.0,1.0,12.0,0.0,0.0,0.0,2,NaN,NaN,0.823333,0.124379,892,3757,4649


In [181]:
# Find the index positions of rows where 'temp' is NaN
# Collect the relevant rows
frames = []

# Find the index positions of rows where 'temp' is NaN
nan_indices = df_bikes[df_bikes['temp'].isna()].index

for idx in nan_indices:
    start = max(idx - 1, 0)
    end = min(idx + 2, len(df_bikes))
    frames.append(df_bikes.iloc[start:end])

# Combine and display
pd.concat(frames)

,instant,dteday,season,yr,mnth,holiday,weekday,workingday,weathersit,temp,atemp,hum,windspeed,casual,registered,cnt
700,701,2012-12-01,4.0,1.0,12.0,0.0,6.0,0.0,2,0.298333,0.316904,0.806667,0.059704,951,4240,5191
701,702,2012-12-02,4.0,1.0,12.0,0.0,0.0,0.0,2,NaN,NaN,0.823333,0.124379,892,3757,4649
702,703,2012-12-03,4.0,1.0,12.0,0.0,1.0,1.0,1,0.452500,0.455796,0.767500,0.082721,555,5679,6234


**Why normalize data**

Note how the yr column is normalized between 0 - 1.0

Normalized data is often more efficient because machine learning weights do not have to adjust for different ranges.

In [182]:
# ----------------------------------------------
# Step 1: Compute the average (mean) temperature
#         using two specific rows: 700 and 702
# ----------------------------------------------

# Take the value of 'temp' from row 700 and row 702,
# add them together and divide by 2 to get the mean.
mean_temp = (df_bikes.iloc[700]['temp'] + df_bikes.iloc[702]['temp']) / 2

# Do the same for 'atemp' (adjusted temperature).
mean_atemp = (df_bikes.iloc[700]['atemp'] + df_bikes.iloc[702]['atemp']) / 2

## helper function to color code
def highlight_mean(val):
    color = 'lightgreen' if val == mean_atemp else ''
    return f'background-color: {color}'

# ----------------------------------------------
# Step 2: Fill missing values (NaNs) in 'temp' and 'atemp' columns
#         with the respective mean values calculated above
# ----------------------------------------------

# For any row where 'temp' is NaN, fill it with mean_temp
df_bikes['temp'] = df_bikes['temp'].fillna(mean_temp)
df_bikes[700:702]['temp']

# Similarly, for any row where 'atemp' is NaN, fill it with mean_atemp
df_bikes['atemp'] = df_bikes['atemp'].fillna(mean_atemp)
# df_bikes[700:703]['atemp']
df_bikes[700:703].style.map(highlight_mean, subset=['atemp'])

,instant,dteday,season,yr,mnth,holiday,weekday,workingday,weathersit,temp,atemp,hum,windspeed,casual,registered,cnt
700,701,2012-12-01,4.000000,1.000000,12.000000,0.000000,6.000000,0.000000,2,0.298333,0.316904,0.806667,0.059704,951,4240,5191
701,702,2012-12-02,4.000000,1.000000,12.000000,0.000000,0.000000,0.000000,2,0.375417,0.386350,0.823333,0.124379,892,3757,4649
702,703,2012-12-03,4.000000,1.000000,12.000000,0.000000,1.000000,1.000000,1,0.452500,0.455796,0.767500,0.082721,555,5679,6234


**Extrapolate dates**

Handling null values for dates.

When real dates are provided, date values (month, day and year) may be extrapolated.

In [183]:
# Convert the 'dteday' column from string (object) type to datetime format.
# This allows for date-based operations like filtering, resampling, and extracting year/month/day.
df_bikes['dteday'] = pd.to_datetime(df_bikes['dteday'])
df_bikes['dteday']

,dteday
0,2011-01-01
1,2011-01-02
2,2011-01-03
3,2011-01-04
4,2011-01-05
...,...
726,2012-12-27
727,2012-12-28
728,2012-12-29
729,2012-12-30


In [184]:
# Import datetime
import datetime as dt

In [185]:
# Extract the month number from the 'dteday' datetime column
# The .dt accessor allows access to datetime properties (like year, month, day, etc.)
# This will return an integer from 1 to 12 for each date in the column
# The result is stored in the column called 'mnth'
df_bikes['mnth'] = df_bikes['dteday'].dt.month

In [186]:
# Show last 5 rows
df_bikes.tail()

,instant,dteday,season,yr,mnth,holiday,weekday,workingday,weathersit,temp,atemp,hum,windspeed,casual,registered,cnt
726,727,2012-12-27,1.0,1.0,12,0.0,4.0,1.0,2,0.254167,0.226642,0.652917,0.350133,247,1867,2114
727,728,2012-12-28,1.0,1.0,12,0.0,5.0,1.0,2,0.253333,0.255046,0.590000,0.155471,644,2451,3095
728,729,2012-12-29,1.0,1.0,12,0.0,6.0,0.0,2,0.253333,0.242400,0.752917,0.124383,159,1182,1341
729,730,2012-12-30,1.0,1.0,12,0.0,0.0,0.0,1,0.255833,0.231700,0.483333,0.350754,364,1432,1796
730,731,2012-12-31,1.0,NaN,12,0.0,1.0,0.0,2,0.215833,0.223487,0.577500,0.154846,439,2290,2729


In [187]:
# Change row 730, column 'yr' to 1.0
df_bikes.loc[730, 'yr'] = 1.0

In [188]:
# Show last 5 rows
df_bikes.tail()

,instant,dteday,season,yr,mnth,holiday,weekday,workingday,weathersit,temp,atemp,hum,windspeed,casual,registered,cnt
726,727,2012-12-27,1.0,1.0,12,0.0,4.0,1.0,2,0.254167,0.226642,0.652917,0.350133,247,1867,2114
727,728,2012-12-28,1.0,1.0,12,0.0,5.0,1.0,2,0.253333,0.255046,0.590000,0.155471,644,2451,3095
728,729,2012-12-29,1.0,1.0,12,0.0,6.0,0.0,2,0.253333,0.242400,0.752917,0.124383,159,1182,1341
729,730,2012-12-30,1.0,1.0,12,0.0,0.0,0.0,1,0.255833,0.231700,0.483333,0.350754,364,1432,1796
730,731,2012-12-31,1.0,1.0,12,0.0,1.0,0.0,2,0.215833,0.223487,0.577500,0.154846,439,2290,2729


**Deleting non-numerical columns**

In [189]:
# Drop 'dteday' column
df_bikes = df_bikes.drop('dteday', axis=1)
df_bikes.head()

,instant,season,yr,mnth,holiday,weekday,workingday,weathersit,temp,atemp,hum,windspeed,casual,registered,cnt
0,1,1.0,0.0,1,0.0,6.0,0.0,2,0.344167,0.363625,0.805833,0.160446,331,654,985
1,2,1.0,0.0,1,0.0,0.0,0.0,2,0.363478,0.353739,0.696087,0.248539,131,670,801
2,3,1.0,0.0,1,0.0,1.0,1.0,1,0.196364,0.189405,0.437273,0.248309,120,1229,1349
3,4,1.0,0.0,1,0.0,2.0,1.0,1,0.200000,0.212122,0.590435,0.160296,108,1454,1562
4,5,1.0,0.0,1,0.0,3.0,1.0,1,0.226957,0.229270,0.436957,0.186900,82,1518,1600


**Predicting regression**

In the bike rentals dataset, df_bikes['cnt'] is the number of bike rentals in a given day. **Predicting** this column would be of great use to a bike rental company. Our problem is to predict the correct number of bike rentals on a given day based on data such as whether this day is a holiday or working day, forecasted temperature, humidity, windspeed, and so on.

**Drop Leakage variables**

### ❓ Why Drop `casual` and `registered` Columns When Predicting `cnt`?

```markdown
According to the dataset, the target variable `cnt` (total bike rentals) is the **sum** of:

- `casual` → number of casual (unregistered) users
- `registered` → number of registered users

That is:
```
cnt = casual + registered
```

Including both `casual` and `registered` as input features would result in a model that **can trivially predict `cnt` with 100% accuracy**, because the answer is literally encoded in the inputs.

### 🧠 Why This Is a Problem:

- In real-world use, we want to **predict `cnt` before the day ends** — before `casual` and `registered` are known.
- These two columns are **leakage variables** — they contain future or outcome-dependent information.
- Keeping them in the input means the model is cheating — it's learning from information it wouldn't have at prediction time.

### ✅ Best Practice:

- Drop `casual` and `registered` from the feature set (`X`) when training the model.
- Train using only those features that would be **known or measurable ahead of time** (e.g., weather, holiday, season, etc.)

```python
X = df_bikes.drop(columns=['cnt', 'casual', 'registered'])
y = df_bikes['cnt']
```

In [190]:
# Drop 'casual', 'registered' columns
df_bikes = df_bikes.drop(['casual', 'registered'], axis=1)
df_bikes.head()

,instant,season,yr,mnth,holiday,weekday,workingday,weathersit,temp,atemp,hum,windspeed,cnt
0,1,1.0,0.0,1,0.0,6.0,0.0,2,0.344167,0.363625,0.805833,0.160446,985
1,2,1.0,0.0,1,0.0,0.0,0.0,2,0.363478,0.353739,0.696087,0.248539,801
2,3,1.0,0.0,1,0.0,1.0,1.0,1,0.196364,0.189405,0.437273,0.248309,1349
3,4,1.0,0.0,1,0.0,2.0,1.0,1,0.200000,0.212122,0.590435,0.160296,1562
4,5,1.0,0.0,1,0.0,3.0,1.0,1,0.226957,0.229270,0.436957,0.186900,1600


**Saving data for future use**

In [191]:
# Export the cleaned DataFrame to a CSV file named 'bike_rentals_cleaned.csv'
# index=False ensures that the row index (0, 1, 2, ...) is not written to the file as an extra column
df_bikes.to_csv('bike_rentals_cleaned.csv', index=False)

**Declaring predictor and target columns**

Here predictor columns are nothing but the input columns and target column is the output column

It's standard practice to group the predictor columns with a capital X, and the target column as a lowercase y

In [192]:
# --------------------------------------------------------
# Split the dataset into features (X) and target (y)
# using positional indexing with .iloc[]
# --------------------------------------------------------

# X: Select all rows (:) and all columns except the last one (:-1)
# Explanation:
# - .iloc[:, :-1] means:
#   - ":" → all rows
#   - ":-1" → all columns from the start up to (but not including) the last column
# This is a common way to isolate the feature columns if the last column is the label
X = df_bikes.iloc[:, :-1]

# y: Select all rows (:) and only the last column (-1)
# Explanation:
# - .iloc[:, -1] means:
#   - ":" → all rows
#   - "-1" → exactly the last column
# This is commonly used when the last column is the target variable (e.g., 'cnt')
y = df_bikes.iloc[:, -1]

**Understanding regression**

### 📌 **Bike Rental Prediction is a Regression Problem**

Predicting the number of bike rentals involves estimating a **non-negative integer** value, which makes it a **regression task** in machine learning — where the target can take on a continuous range of values.

One of the most common algorithms for this is **linear regression**, which models the target as a weighted sum of input features (predictors). These weights (also called **coefficients**) are learned using an optimization technique like **gradient descent**, which minimizes prediction error. The output of linear regression can be **any real number (including negative values)**, even though the actual rental count is always a non-negative integer.


**How to Handle negative predictions?**

> 👉 **Linear regression predictions can be *any real number*, including negative values.**

---

### ❓ But Why Is That a Problem?

In many real-world problems — like predicting:

- **Bike rentals**
- **Housing prices**
- **Population counts**
- **Sales volume**

…the **target variable can’t be negative**. But since linear regression is a **mathematical model**, it doesn't "know" these domain-specific rules — it simply finds the **best-fit line**.

So if the inputs (features) push the prediction far enough, it **can return negative values**.

---

### 🛠 How to Handle This:

| Approach | Description |
|----------|-------------|
| **Post-processing** | Clamp predictions with `max(0, prediction)` to avoid negative outputs. |
| **Use models with non-negative constraints** | Some versions of regression (e.g., `sklearn.linear_model.Lasso` or `Ridge` with constraints) can limit weights. |
| **Log-transform the target** | Model `log(y)` instead of `y`, then apply `exp()` to get the final output. This ensures positive predictions. |
| **Use tree-based models** | Models like XGBoost or RandomForest naturally avoid unreasonable negatives in most regression tasks. |

**Predicting total number of bike rentals for each day**

In [193]:
# --------------------------------------------------------
# Step 1: Import train_test_split function
# --------------------------------------------------------
# This utility is used to split the dataset into training and test subsets
# It helps evaluate how well the model generalizes to unseen data
from sklearn.model_selection import train_test_split

# --------------------------------------------------------
# Step 2: Import LinearRegression model class
# --------------------------------------------------------
# This is scikit-learn’s implementation of Ordinary Least Squares (OLS) linear regression
from sklearn.linear_model import LinearRegression

# --------------------------------------------------------
# Step 3: Split the dataset into training and test sets
# --------------------------------------------------------
# - X: feature matrix (input variables)
# - y: target vector (output variable)
# - test_size is not explicitly set, so default is 0.25 (25% test, 75% train)
# - random_state=2 ensures reproducibility (same split each time)
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=2)

 Whenever you see random_state=2, this means that you are choosing the seed of a pseudo-random number generator to ensure reproducible results.

In [194]:
# --------------------------------------------------------
# Step: Silence warnings in the notebook or script output
# --------------------------------------------------------

# Import the warnings module — part of Python's standard library
# It lets you control how warning messages are handled
import warnings

# Suppress all warning messages globally
# This is useful to hide non-critical warnings (e.g., deprecation or convergence warnings)
# Typically done for cleaner notebook outputs or demos
warnings.filterwarnings('ignore')

In [195]:
# --------------------------------------------
# Step 1: Initialize the Linear Regression model
# --------------------------------------------
# Create an instance of scikit-learn's LinearRegression class
# This model will learn the relationship between features (X) and target (y)
lin_reg = LinearRegression()

# --------------------------------------------
# Step 2: Train the model on the training set
# --------------------------------------------
# Fit the linear regression model using the training data
# X_train: input features, y_train: corresponding labels
lin_reg.fit(X_train, y_train)

# --------------------------------------------
# Step 3: Predict on the test set
# --------------------------------------------
# Use the trained model to predict the target values for the test features (X_test)
y_pred = lin_reg.predict(X_test)

# --------------------------------------------
# Step 4: Evaluate the model using RMSE
# --------------------------------------------
# Import the mean_squared_error function from scikit-learn
from sklearn.metrics import mean_squared_error

# Import numpy for square root computation
import numpy as np

# Calculate Mean Squared Error (MSE) between actual and predicted test labels
mse = mean_squared_error(y_test, y_pred)

# Take square root of MSE to get Root Mean Squared Error (RMSE)
# RMSE gives a measure of how far predictions are from actual values on average
rmse = np.sqrt(mse)

# --------------------------------------------
# Step 5: Display the RMSE
# --------------------------------------------
# Print the RMSE rounded to 2 decimal places
print("RMSE: %0.2f" % (rmse))

# --------------------------------------------
# Step 6: Display R2 score
# --------------------------------------------
from sklearn.metrics import r2_score
r2 = r2_score(y_test, y_pred)
print("R² Score: %0.4f" % r2)

RMSE: 898.21
R² Score: 0.7933


**Evaluation:** How to determine if RMSE score of 898.21 good or bad?

In [196]:
# Display bike rental stats
df_bikes['cnt'].describe()

,cnt
count,731.000000
mean,4504.348837
std,1937.211452
min,22.000000
25%,3152.000000
50%,4548.000000
75%,5956.000000
max,8714.000000


## Is RMSE of 898.21 good or bad? 👇

---

## 🎯 What You're Trying to Do

You're using **Linear Regression** to predict `cnt` — the total number of bike rentals for each day. After training your model, you got this:

```text
RMSE = 898.21
```

You also checked the statistics of the actual `cnt` values:

| Metric      | Value     |
|-------------|-----------|
| Mean        | 4504      |
| Std Dev     | 1937      |
| Min / Max   | 22 → 8714 |
| 25% / 75%   | 3152 → 5956 |

---

## 🤔 So What Does RMSE = 898 Mean?

> RMSE tells you, **on average**, how far off your model’s predictions are — **in the same unit** as the target (bike counts).

So:
```
Your model is off by ~898 rentals per day on average.
```

---

## 🧠 Is That Good or Bad?

Let’s compare it to some context from your data:

### 🔍 Relative to the **mean (average)**:
- Mean bike count = 4504  
- RMSE = 898  
- That’s roughly **20% of the average**:
  $$
  \frac{898}{4504} \approx 0.20
  $$

That means: **your model is about 80% accurate** (roughly speaking) in terms of scale.

---

### ✅ Why This Is Decent for a Beginner Model

- You’re using **plain Linear Regression** — a simple model.
- Real-world data is **noisy** (weather, holidays, weekends, etc.).
- Getting within ±900 of a value that can range from **22 to 8714** is **not bad at all** for a first model.

---

## 📊 Visual Intuition

Imagine a daily bike rental prediction:
```
True value: 4600
Your model says: 3700 → Error = -900
```

That kind of variation is acceptable — especially if you’re predicting for planning or analytics.

---

## 📌 TL;DR for Beginners

| Metric | Value | What It Tells You |
|--------|-------|-------------------|
| RMSE   | 898   | You're off by ~898 bikes on average |
| Mean   | 4504  | Typical daily rental count |
| Verdict | ✅ Decent! | Especially for a first model using linear regression |

---

## 📌 How to reduce that RMSE further?

maybe using XGBoost or feature engineering?

---
# **Using XGBoost**

In [197]:
# ----------------------------------------------------
# Step 1: Import the XGBoost Regressor
# ----------------------------------------------------
# XGBRegressor is a high-performance, tree-based model optimized for speed and accuracy
from xgboost import XGBRegressor

# ----------------------------------------------------
# Step 2: Instantiate the model
# ----------------------------------------------------
# Create an instance of XGBRegressor using default hyperparameters
# You can customize it later with parameters like n_estimators, learning_rate, etc.
xg_reg = XGBRegressor()

# ----------------------------------------------------
# Step 3: Train the model on the training set
# ----------------------------------------------------
# Fit the model to the training data (X_train: features, y_train: target)
xg_reg.fit(X_train, y_train)

# ----------------------------------------------------
# Step 4: Predict on the test set
# ----------------------------------------------------
# Use the trained model to predict the target values for X_test
# The output y_pred contains predicted values corresponding to y_test
y_pred = xg_reg.predict(X_test)

# ----------------------------------------------------
# Step 5: Evaluate the model using RMSE
# ----------------------------------------------------
# Calculate Mean Squared Error between actual and predicted values
mse = mean_squared_error(y_test, y_pred)

# Take the square root of MSE to compute Root Mean Squared Error (RMSE)
# RMSE gives a more interpretable error metric in the original unit of the target variable
rmse = np.sqrt(mse)

# ----------------------------------------------------
# Step 6: Display the RMSE result
# ----------------------------------------------------
# Print the RMSE, rounded to 2 decimal places, to summarize prediction error
print("RMSE: %0.2f" % (rmse))

# --------------------------------------------
# Step 6: Display R2 score
# --------------------------------------------
from sklearn.metrics import r2_score
r2 = r2_score(y_test, y_pred)
print("R² Score: %0.4f" % r2)

RMSE: 693.21
R² Score: 0.8769


## ***SIDEBAR:***

Let's compare your **Linear Regression (RMSE: 898.21)** and **XGBoost Regression (RMSE: 693.21)** against the context of your target variable `cnt`.

---

## 🎯 What Are We Comparing?

| Model                | **RMSE** |
|----------------------|----------|
| **Linear Regression**| 898.21   |
| **XGBoost Regressor**| 693.21   |

Your goal is to **predict daily bike rental counts** (target = `cnt`), and here’s what the actual values look like:

### 📊 `df_bikes['cnt'].describe()`:

| Statistic | Value     |
|-----------|-----------|
| **count** | 731       |
| **mean**  | 4504      |
| **std**   | 1937      |
| **min**   | 22        |
| **25%**   | 3152      |
| **50%**   | 4548      |
| **75%**   | 5956      |
| **max**   | 8714      |

---

## 📈 RMSE Comparison in Context

### 📌 Linear Regression RMSE: **898**
- Roughly **20% of the average value (4504)**.
- Acceptable, but still off by a significant margin.

### ✅ XGBoost RMSE: **693**
- Roughly **15.4% of the average value**:
  $$
  \frac{693.21}{4504.34} \approx 0.154
  $$
- That’s **~23% lower error** compared to linear regression.
- Much tighter predictions — especially important if you're making decisions based on this.

---

## 🧠 Interpretation for Beginners

| Model       | Avg Error (RMSE) | Performance | In Plain English |
|-------------|------------------|-------------|------------------|
| Linear      | 898 bikes         | Baseline    | On average, off by ~900 rentals per day |
| XGBoost     | 693 bikes         | ✅ Better!  | On average, off by ~693 rentals per day — tighter, more accurate |

🟢 **XGBoost is doing a better job capturing the nonlinear patterns** — like the effect of holidays, temperature, or weekday vs weekend — that linear regression may miss.

---

## 📐 Now Let’s Talk About R² Score

While RMSE tells us **how wrong our predictions are (in actual units)**, the **R² Score** tells us **how useful our model is overall**.

### ✅ R² Score (Coefficient of Determination)

| Model              | **R² Score** |
|--------------------|-------------|
| Linear Regression  | 0.7933      |
| XGBoost Regressor  | 0.8769      |

### 💡 What Does That Mean?

> R² Score answers: **"How much of the variation in daily bike rentals can the model explain?"**

- **R² = 1** → perfect predictions  
- **R² = 0** → model is no better than just guessing the average  
- **R² < 0** → model is worse than guessing!

---

### 🧠 Beginner Intuition:

| Model       | R² Score | Meaning |
|-------------|----------|---------|
| Linear      | 0.7933   | Explains **79%** of the variability in `cnt` |
| XGBoost     | 0.8769   | Explains **88%** of the variability — ✅ better |

So not only is XGBoost making smaller errors (lower RMSE), it’s also **understanding more of what influences bike rentals**.

---

## 🛠 What Makes XGBoost Better?

- Handles **nonlinear relationships** naturally
- Models **feature interactions** automatically
- Uses **regularization** to reduce overfitting
- Handles **missing values** and **irrelevant features** intelligently

---

## ✅ TL;DR Summary

> **XGBoost improves RMSE by ~205 rentals/day** and boosts R² by nearly **8.3 percentage points** — a meaningful gain on both accuracy and explainability.

| Metric       | Linear Reg | XGBoost   |
|--------------|------------|-----------|
| **RMSE**     | 898.21     | 693.21    |
| **% of Mean**| ~20%       | ~15.4%    |
| **R² Score** | 0.7933     | 0.8769    |
| **Verdict**  | Good start | ✅ Better predictions and understanding |

---

## 🧪 What You Can Try Next

### 📌 Improve the Model:
- **Hyperparameter tuning** (grid search / random search)
- **Feature engineering** (e.g., extract `weekday`, `is_holiday`, `season_avg`)
- **Log transform the target** to reduce impact of extreme values:
  ```python
  y_log = np.log1p(y_train)
  model.fit(X_train, y_log)
  y_pred = np.expm1(model.predict(X_test))
  ```

# ***💡 Tip:***
- XGBoost often performs better than linear models on nonlinear or complex datasets.
- You can further improve performance by tuning hyperparameters like `max_depth`, `learning_rate`, `n_estimators`, etc.

## **Cross-validation**

**Cross-validation with linear regression**

In [198]:
# -----------------------------------------------------------
# Step 1: Import the cross_val_score utility from sklearn
# -----------------------------------------------------------
# cross_val_score allows you to easily perform k-fold cross-validation
from sklearn.model_selection import cross_val_score

# -----------------------------------------------------------
# Step 2: Instantiate the regression model (Linear Regression)
# -----------------------------------------------------------
# This creates the model object we'll pass to cross_val_score
model = LinearRegression()

# -----------------------------------------------------------
# Step 3: Perform 10-fold cross-validation
# -----------------------------------------------------------
# - model: the regression algorithm to evaluate
# - X, y: input features and target variable
# - scoring='neg_mean_squared_error': sklearn returns negative MSE (so we need to invert it later)
# - cv=10: use 10-fold cross-validation
scores = cross_val_score(model, X, y, scoring='neg_mean_squared_error', cv=10)

# -----------------------------------------------------------
# Step 4: Convert negative MSE to positive RMSE
# -----------------------------------------------------------
# -scores converts to positive MSE
# np.sqrt() gives us RMSE (Root Mean Squared Error)
rmse = np.sqrt(-scores)

# -----------------------------------------------------------
# Step 5: Print RMSE for each of the 10 folds
# -----------------------------------------------------------
# Round each RMSE value to 2 decimal places for readability
print('Reg rmse:', np.round(rmse, 2))

# -----------------------------------------------------------
# Step 6: Print the average RMSE across all folds
# -----------------------------------------------------------
# This gives a more stable estimate of model performance
print('RMSE mean: %0.2f' % (rmse.mean()))

Reg rmse: [ 504.01  840.55 1140.88  728.39  640.2   969.95 1133.45 1252.85 1084.64
 1425.33]
RMSE mean: 972.02


***Tip***

Why **scoring='neg_mean_squared_error'**?

Scikit-learn is designed to select the highest score when training models. This works well for accuracy, but not for errors when the lowest is best. By taking the negative of each mean squared error, the lowest ends up being the highest. This is compensated for later with rmse = np.sqrt(-scores), so the final results are positive.

## ✅ *SIIDEBAR: Explaining the Output*

```
Reg rmse: [ 504.01  840.55 1140.88  728.39  640.2   969.95 1133.45 1252.85 1084.64 1425.33]
RMSE mean: 972.02
```

---

## 🧠 What This Means

### 🔢 1. **These are the RMSE scores from each fold**

You performed **10-fold cross-validation**, so:

- Your dataset was split into **10 equal parts (folds)**.
- The model was trained and tested **10 times**, each time on a different fold.
- The numbers in `Reg rmse:` represent the **RMSE (Root Mean Squared Error)** from each of those 10 test sets.

| Fold | RMSE      |
|------|-----------|
| 1    | 504.01    |
| 2    | 840.55    |
| 3    | 1140.88   |
| 4    | 728.39    |
| 5    | 640.20    |
| 6    | 969.95    |
| 7    | 1133.45   |
| 8    | 1252.85   |
| 9    | 1084.64   |
| 10   | 1425.33   |

---

### 📈 2. **RMSE mean: 972.02**

- This is the **average error** across all 10 folds.
- It tells you that **on average, your model’s predictions are off by ~972 bike rentals**.

---

## 🔍 How to Interpret

| Metric       | Value    | Meaning |
|--------------|----------|---------|
| **Low RMSE (e.g., 504.01)** | Some folds had easier-to-predict test sets |
| **High RMSE (e.g., 1425.33)** | Some test sets were more difficult (e.g., holidays, outliers) |
| **Mean RMSE = 972.02** | More reliable estimate of overall model performance |

---

## ⚠️ The Variation Tells You...

- There’s **significant fluctuation** in RMSE across folds (e.g., 504 → 1425).
- This might mean:
  - Your data has **outliers**
  - Some time periods are harder to predict
  - Model performance varies based on the test set used

---

## ✅ Next Steps to Improve

- Try a **more expressive model** like `RandomForestRegressor()` or `XGBRegressor()`
- Add **seasonal features** like `month`, `weekday`, or `is_holiday`
- Consider **log-transforming the target** if there are big spikes in `cnt`

**Cross-validation with XGBoost**

In [199]:
# -----------------------------------------------------------
# Step 1: Instantiate the XGBoost Regressor
# -----------------------------------------------------------
# - We set the objective to "reg:squarederror" for regression tasks
# - This tells XGBoost to use squared error as the loss function
model = XGBRegressor(objective="reg:squarederror")

# -----------------------------------------------------------
# Step 2: Perform 10-fold cross-validation
# -----------------------------------------------------------
# - model: the regression algorithm to evaluate
# - X, y: input features and target variable
# - scoring='neg_mean_squared_error': returns negative MSE (standard in sklearn)
# - cv=10: perform 10 different train/test splits (folds)
scores = cross_val_score(model, X, y, scoring='neg_mean_squared_error', cv=10)

# -----------------------------------------------------------
# Step 3: Convert negative MSE to positive RMSE
# -----------------------------------------------------------
# -scores gives us positive MSE
# np.sqrt() converts it to RMSE (Root Mean Squared Error)
rmse = np.sqrt(-scores)

# -----------------------------------------------------------
# Step 4: Print RMSE for each fold
# -----------------------------------------------------------
# This shows the prediction error (in original units) for each test fold
print('Reg rmse:', np.round(rmse, 2))

# -----------------------------------------------------------
# Step 5: Print the average RMSE across all folds
# -----------------------------------------------------------
# This is a more stable estimate of model performance than any single test split
print('RMSE mean: %0.2f' % (rmse.mean()))

Reg rmse: [ 869.03  902.85  556.63  685.42  890.73  930.6   890.65  879.29  747.78
 2067.64]
RMSE mean: 942.06


**XGBRegressor wins again, besting linear regression by about 10%.**

***SIDEBAR***

## ✅ Your Output

```text
Reg rmse: [ 869.03  902.85  556.63  685.42  890.73  930.6   890.65  879.29  747.78 2067.64]
RMSE mean: 942.06
```

---

## 🔍 What This Means

### 🔢 1. **Each value in `Reg rmse` is the RMSE from one of the 10 folds**
- You performed **10-fold cross-validation**, meaning:
  - The data was split into 10 parts.
  - In each run, 1 part was used as the test set, the other 9 parts as the training set.
  - You repeated this 10 times, and got 10 **RMSE values**.

| Fold | RMSE    | Comment                         |
|------|---------|----------------------------------|
| 1    | 869.03  | Fairly low error                |
| 2    | 902.85  | Similar to fold 1               |
| 3    | 556.63  | 🔽 Very low error – easy-to-predict test set |
| 4    | 685.42  | Decent                          |
| 5    | 890.73  | Similar to fold 1 and 2         |
| 6    | 930.60  | Slightly higher                 |
| 7    | 890.65  | Consistent                      |
| 8    | 879.29  | Consistent                      |
| 9    | 747.78  | Lower than average              |
| 10   | 2067.64 | 🔺 Outlier – possibly holiday or event days, or a small, hard-to-predict subset |

---

### 📈 2. **Mean RMSE = 942.06**

- This is the **average of all 10 RMSEs**.
- It tells you that, **on average, your XGBoost model is off by ~942 bike rentals per day**.

---

## 🧠 How to Interpret This

| Insight | Explanation |
|--------|-------------|
| ✅ **Good performance** in most folds | 9 out of 10 folds had RMSE between ~550–930 |
| ⚠️ **One very high error** (2067.64) | Suggests that fold 10 had very hard-to-predict data |
| 🧮 **RMSE mean = 942.06** | Slightly better or similar to Linear Regression depending on your earlier scores |

---

## 📌 Why is Fold 10 so high?

This fold may include:
- **Unusual days** (e.g., holidays, weather spikes)
- **Rare events** not well represented in training
- A **small sample size** with noisy target values

---

## ✅ What You Can Do Next

| Step | Reason |
|------|--------|
| **Inspect Fold 10** | Understand what made it difficult to predict |
| **Try stratified k-folds** (if target has skew) | To ensure consistent target distribution |
| **Add new features** | e.g., `is_weekend`, `is_holiday`, `temp_range`, etc. |
| **Log transform the target** | If target (`cnt`) is highly skewed |
| **Try outlier removal or smoothing** | For more robust training |


# Predicting Classification

**Data loading**

In [200]:
# Upload Census dataset (adult) from UCI Machine Learning Repository
df_census = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.data')

# Display first 5 rows
df_census.head()

,39,State-gov,77516,Bachelors,13,Never-married,Adm-clerical,Not-in-family,White,Male,2174,0,40,United-States,<=50K
0,50,Self-emp-not-inc,83311,Bachelors,13,Married-civ-spouse,Exec-managerial,Husband,White,Male,0,0,13,United-States,<=50K
1,38,Private,215646,HS-grad,9,Divorced,Handlers-cleaners,Not-in-family,White,Male,0,0,40,United-States,<=50K
2,53,Private,234721,11th,7,Married-civ-spouse,Handlers-cleaners,Husband,Black,Male,0,0,40,United-States,<=50K
3,28,Private,338409,Bachelors,13,Married-civ-spouse,Prof-specialty,Wife,Black,Female,0,0,40,Cuba,<=50K
4,37,Private,284582,Masters,14,Married-civ-spouse,Exec-managerial,Wife,White,Female,0,0,40,United-States,<=50K


In [201]:
# Upload Census dataset with no header
df_census = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.data', header=None)

# Display first 5 rows
df_census.head()

,0,1,2,3,4,5,6,7,8,9,10,11,12,13,14
0,39,State-gov,77516,Bachelors,13,Never-married,Adm-clerical,Not-in-family,White,Male,2174,0,40,United-States,<=50K
1,50,Self-emp-not-inc,83311,Bachelors,13,Married-civ-spouse,Exec-managerial,Husband,White,Male,0,0,13,United-States,<=50K
2,38,Private,215646,HS-grad,9,Divorced,Handlers-cleaners,Not-in-family,White,Male,0,0,40,United-States,<=50K
3,53,Private,234721,11th,7,Married-civ-spouse,Handlers-cleaners,Husband,Black,Male,0,0,40,United-States,<=50K
4,28,Private,338409,Bachelors,13,Married-civ-spouse,Prof-specialty,Wife,Black,Female,0,0,40,Cuba,<=50K


In [202]:
# Define df_census columns
# These names can be found here - (https://archive.ics.uci.edu/ml/datasets/Census+Income) under Attribute Information
df_census.columns = ['age', 'workclass', 'fnlwgt', 'education', 'education-num', 'marital-status', 'occupation',
                  'relationship', 'race', 'sex', 'capital-gain', 'capital-loss', 'hours-per-week', 'native-country',
                   'income']

# Display first 5 rows
df_census.head()

,age,workclass,fnlwgt,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loss,hours-per-week,native-country,income
0,39,State-gov,77516,Bachelors,13,Never-married,Adm-clerical,Not-in-family,White,Male,2174,0,40,United-States,<=50K
1,50,Self-emp-not-inc,83311,Bachelors,13,Married-civ-spouse,Exec-managerial,Husband,White,Male,0,0,13,United-States,<=50K
2,38,Private,215646,HS-grad,9,Divorced,Handlers-cleaners,Not-in-family,White,Male,0,0,40,United-States,<=50K
3,53,Private,234721,11th,7,Married-civ-spouse,Handlers-cleaners,Husband,Black,Male,0,0,40,United-States,<=50K
4,28,Private,338409,Bachelors,13,Married-civ-spouse,Prof-specialty,Wife,Black,Female,0,0,40,Cuba,<=50K


**Data wrangling**

The dataset must be preprocessed. When testing new algorithms, it's essential to have all numerical columns with no null values.

**Null values**

A great way to check null values is to look at the DataFrame .info() method

In [203]:
# Display df_census info
df_census.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             32561 non-null  int64 
 1   workclass       32561 non-null  object
 2   fnlwgt          32561 non-null  int64 
 3   education       32561 non-null  object
 4   education-num   32561 non-null  int64 
 5   marital-status  32561 non-null  object
 6   occupation      32561 non-null  object
 7   relationship    32561 non-null  object
 8   race            32561 non-null  object
 9   sex             32561 non-null  object
 10  capital-gain    32561 non-null  int64 
 11  capital-loss    32561 non-null  int64 
 12  hours-per-week  32561 non-null  int64 
 13  native-country  32561 non-null  object
 14  income          32561 non-null  object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB


Since all columns have the same number of non-null rows, we can infer that there are no null values.

**Non-numerical columns**

All columns of the dtype object must be transformed into numerical columns

In [204]:
# Drop education column
# The education_num column is a numerical conversion of 'education'. Since the information is the same, the 'education' column may be deleted
df_census = df_census.drop(['education'], axis=1)

In [205]:
df_census.head()

,age,workclass,fnlwgt,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loss,hours-per-week,native-country,income
0,39,State-gov,77516,13,Never-married,Adm-clerical,Not-in-family,White,Male,2174,0,40,United-States,<=50K
1,50,Self-emp-not-inc,83311,13,Married-civ-spouse,Exec-managerial,Husband,White,Male,0,0,13,United-States,<=50K
2,38,Private,215646,9,Divorced,Handlers-cleaners,Not-in-family,White,Male,0,0,40,United-States,<=50K
3,53,Private,234721,7,Married-civ-spouse,Handlers-cleaners,Husband,Black,Male,0,0,40,United-States,<=50K
4,28,Private,338409,13,Married-civ-spouse,Prof-specialty,Wife,Black,Female,0,0,40,Cuba,<=50K


In [206]:
# -------------------------------------------------------------
# Step 1: Convert non-numeric (categorical) columns to numeric
# -------------------------------------------------------------
# pd.get_dummies() automatically:
# - Detects categorical columns (e.g., object or category dtype)
# - Replaces them with one-hot encoded numeric columns
#   For example:
#     A column like 'Gender' with values ['Male', 'Female']
#     becomes two columns: 'Gender_Male', 'Gender_Female'
# - Each row gets a 1 in the column matching its category, 0 elsewhere
# - All other columns (numeric) are preserved as-is

df_census = pd.get_dummies(df_census)

# -------------------------------------------------------------
# Step 2: Display the first 5 rows of the transformed DataFrame
# -------------------------------------------------------------
# This helps verify:
# - Which columns were converted
# - The shape and structure after encoding
df_census.head()

,age,fnlwgt,education-num,capital-gain,capital-loss,hours-per-week,workclass_ ?,workclass_ Federal-gov,workclass_ Local-gov,workclass_ Never-worked,...,native-country_ Scotland,native-country_ South,native-country_ Taiwan,native-country_ Thailand,native-country_ Trinadad&Tobago,native-country_ United-States,native-country_ Vietnam,native-country_ Yugoslavia,income_ <=50K,income_ >50K
0,39,77516,13,2174,0,40,False,False,False,False,...,False,False,False,False,False,True,False,False,True,False
1,50,83311,13,0,0,13,False,False,False,False,...,False,False,False,False,False,True,False,False,True,False
2,38,215646,9,0,0,40,False,False,False,False,...,False,False,False,False,False,True,False,False,True,False
3,53,234721,7,0,0,40,False,False,False,False,...,False,False,False,False,False,True,False,False,True,False
4,28,338409,13,0,0,40,False,False,False,False,...,False,False,False,False,False,False,False,False,True,False


In [207]:
print(df_census.columns)

Index(['age', 'fnlwgt', 'education-num', 'capital-gain', 'capital-loss',
       'hours-per-week', 'workclass_ ?', 'workclass_ Federal-gov',
       'workclass_ Local-gov', 'workclass_ Never-worked', 'workclass_ Private',
       'workclass_ Self-emp-inc', 'workclass_ Self-emp-not-inc',
       'workclass_ State-gov', 'workclass_ Without-pay',
       'marital-status_ Divorced', 'marital-status_ Married-AF-spouse',
       'marital-status_ Married-civ-spouse',
       'marital-status_ Married-spouse-absent',
       'marital-status_ Never-married', 'marital-status_ Separated',
       'marital-status_ Widowed', 'occupation_ ?', 'occupation_ Adm-clerical',
       'occupation_ Armed-Forces', 'occupation_ Craft-repair',
       'occupation_ Exec-managerial', 'occupation_ Farming-fishing',
       'occupation_ Handlers-cleaners', 'occupation_ Machine-op-inspct',
       'occupation_ Other-service', 'occupation_ Priv-house-serv',
       'occupation_ Prof-specialty', 'occupation_ Protective-serv',
     

**Target and predictor columns**

The target column is whether or not someone makes 50K.

In [45]:
# Drop column 'income_ <=50K'
# This is a multicollinearity issue
# For more details - 1.8-Study Guide: Preparing the Target Column for Classification — income > 50K.md
df_census = df_census.drop('income_ <=50K', axis=1)

In [46]:
# ------------------------------------------------------------
# Step: Split the dataset into features (X) and target (y)
# ------------------------------------------------------------

# X = all rows, and all columns except the last one (features)
# df.iloc[row_slice, column_slice]
# - iloc[:,:-1] means: for every row (:), select all columns except the last one (:-1)
# - This works assuming the last column is the target (e.g., 'income_>50K')
X = df_census.iloc[:, :-1]

# y = all rows, and only the last column (target variable)
# df.iloc[row_slice, column_slice]
# - iloc[:,-1] means: for every row (:), select only the last column (-1)
# - This gives us the binary target column we kept (e.g., 'income_>50K')
y = df_census.iloc[:, -1]

**Logistic regression**

In [47]:
# Import Logistic Regression
from sklearn.linear_model import LogisticRegression

In [48]:
# -------------------------------------------------------------
# Step 1: Import cross_val_score from scikit-learn
# -------------------------------------------------------------
# cross_val_score handles k-fold cross-validation behind the scenes
from sklearn.model_selection import cross_val_score

# -------------------------------------------------------------
# Step 2: Define a function to evaluate a classifier using CV
# -------------------------------------------------------------
# This function takes:
# - classifier: the model you want to evaluate (e.g., LogisticRegression)
# - num_splits: how many folds (default is 10 for 10-fold CV)
def cross_val(classifier, num_splits=10):

    # Step 2a: Initialize the classifier model
    model = classifier  # e.g., LogisticRegression()

    # Step 2b: Perform cross-validation and return accuracy scores
    # - cv=num_splits splits the data into k folds
    # - The model is trained and tested k times on different subsets
    scores = cross_val_score(model, X, y, cv=num_splits)

    # Step 2c: Print accuracy from each fold (rounded to 2 decimals)
    print('Accuracy:', np.round(scores, 2))

    # Step 2d: Print the average accuracy across all folds
    print('Accuracy mean: %0.2f' % (scores.mean()))



In [49]:
# -------------------------------------------------------------
# Step 3: Call the function with LogisticRegression as classifier
# -------------------------------------------------------------
cross_val(LogisticRegression())

Accuracy: [0.8  0.8  0.79 0.8  0.79 0.81 0.79 0.79 0.8  0.8 ]
Accuracy mean: 0.80


### ❓ Where Is the **Sigmoid Function** Used?

You don’t see it directly in the code, **but it’s built into the LogisticRegression model**:

```python
LogisticRegression()
```

Internally, during:
- **Training**: the model fits weights using **log loss (cross-entropy)**, which is based on the sigmoid function
- **Prediction**: it computes \( z = w^T x + b \), and applies:

$$
\sigma(z) = \frac{1}{1 + e^{-z}}
$$

This sigmoid output gives the **probability of class 1**, which is then used to classify based on a threshold (usually 0.5).

---

### ✅ Example Output Interpretation:

```python
Accuracy: [0.8  0.8  0.79 0.8  0.79 0.81 0.79 0.79 0.8  0.8 ]
Accuracy mean: 0.80
```

This means:
- Across 10 folds, the model was ~80% accurate in each
- The average accuracy is **0.80 (80%)**, which is quite stable and decent


**The XGBoost classifier**

In [50]:
# Import XGBoost Classifier
from xgboost import XGBClassifier

In [51]:
# -------------------------------------------------------------
# Use cross_val function to evaluate XGBoost Classifier
# -------------------------------------------------------------
# XGBClassifier is a powerful tree-based ensemble classifier from the XGBoost library
# - n_estimators=5 means the model will build 5 boosted decision trees
# - Each tree corrects the errors of the previous one (gradient boosting)

# Compared to LogisticRegression:
# - LogisticRegression is a **linear model** that uses the **sigmoid function**
#   to map the linear combination of features to a probability
# - XGBClassifier is a **nonlinear model** based on **decision trees**
#   and uses gradient boosting to improve performance iteratively
# - LogisticRegression has a simple decision boundary (linear)
# - XGBClassifier can learn complex, nonlinear boundaries and feature interactions

# This call performs 10-fold cross-validation by default
# and prints the accuracy per fold and the mean accuracy
cross_val(XGBClassifier(n_estimators=5))


Accuracy: [0.85 0.86 0.86 0.85 0.86 0.86 0.86 0.87 0.86 0.86]
Accuracy mean: 0.86


# **SIDEBAR**

Let’s compare and contrast the **cross-validation results** of your two classifiers: `LogisticRegression()` vs. `XGBClassifier(n_estimators=5)` — in terms of accuracy, consistency, and what it says about model performance.

---

## 🔍 Cross-Validation Output Comparison

### 1️⃣ Logistic Regression

```
Accuracy:    [0.80, 0.80, 0.79, 0.80, 0.79, 0.81, 0.79, 0.79, 0.80, 0.80]
Mean Accuracy: 0.80
```

### 2️⃣ XGBoost Classifier (n_estimators=5)

```
Accuracy:    [0.85, 0.86, 0.86, 0.85, 0.86, 0.86, 0.86, 0.87, 0.86, 0.86]
Mean Accuracy: 0.86
```

---

## 📊 Interpretation

| Aspect                      | `LogisticRegression`     | `XGBClassifier`              |
|-----------------------------|---------------------------|------------------------------|
| **Mean Accuracy**           | **0.80**                  | ✅ **0.86** (higher)         |
| **Best Fold Accuracy**      | 0.81                      | 0.87                         |
| **Worst Fold Accuracy**     | 0.79                      | 0.85                         |
| **Variance Between Folds**  | Slight (~0.02 spread)     | Even tighter (~0.02 spread) |
| **Consistency**             | ✅ High                   | ✅✅ Very high                |

---

## ✅ Key Takeaways

### 📈 Accuracy
- **XGBoost outperforms Logistic Regression** by **6 percentage points on average** (86% vs 80%).
- This suggests XGBoost is capturing **nonlinear patterns** and **interactions** in your data that Logistic Regression cannot.

### 🔁 Consistency Across Folds
- Both models are **fairly stable**, but **XGBoost is even more consistent**, with all scores tightly clustered around 86%.

### ⚙️ Why Is XGBoost Better?
| Reason | Explanation |
|--------|-------------|
| **Nonlinear capability** | XGBoost handles nonlinear relationships natively (via trees) |
| **Boosting** | Each tree improves upon the previous one’s errors |
| **Automatic feature interactions** | Unlike LogisticRegression, XGBoost **learns** feature combinations |
| **Better for tabular data** | Structured datasets (like census, Titanic) often benefit from XGBoost |

---

## 📌 Final Verdict

| Model               | Avg Accuracy | Verdict                 |
|---------------------|--------------|--------------------------|
| Logistic Regression | 0.80         | 🟡 Good baseline         |
| XGBoost Classifier  | 0.86         | ✅ Better performance & stability |

> ✅ **Use XGBoost when accuracy and robustness are more critical**, especially on structured/tabular datasets.

Would you like to add **ROC AUC**, **F1-score**, or visualize both model performances on a chart next?

### ✅ Key Differences: LogisticRegression vs XGBClassifier

| Feature                  | `LogisticRegression()`                   | `XGBClassifier(n_estimators=5)`               |
|--------------------------|------------------------------------------|----------------------------------------------|
| Model type               | Linear (parametric)                      | Tree-based ensemble (non-parametric)         |
| Output mechanism         | Sigmoid activation over linear weights   | Majority vote over boosted decision trees    |
| Handles nonlinearity     | ❌ No                                     | ✅ Yes                                        |
| Feature interactions     | ❌ Must be engineered manually            | ✅ Learned automatically                     |
| Regularization           | ✅ L1/L2                                  | ✅ Built-in (e.g., gamma, lambda, alpha)     |
| Performance on complex data | Often limited                         | ✅ Usually better on structured/tabular data |

---

Let me know if you'd like to expand `cross_val()` to log precision, recall, or ROC AUC across folds too!